### PR TITLE
NEW support selllist in the  module builder

### PR DESCRIPTION
**Pull request by delcroip**
# New support selllist in the  module builder

### CODE REVIEW

**Comment by delcroip** [comment]
I basically copy/pasted the code from extrafileds

**Comment by delcroip** [comment]
here the select_sellist:

function select_sellist($sqlarray=array('table'=> 'user','keyfield'=> 'rowid','fields'=>'firstname,lastname', 'join' => '', 'where'=>'','tail'=>''),
                        $htmlarray=array('name'=> 'HTMLSellist','class'=>'','otherparam'=>'','$ajaxNbChar'=>'','separator'=> ' '),
                        $selected='',
                        $addtionnalChoices=array('NULL'=>'NULL')){
        global $conf,$langs,$db;
     $ajax=$conf->use_javascript_ajax ;
    if( !isset($sqlarray['table'])|| !isset($sqlarray['keyfield'])||!isset($sqlarray['fields']) || !isset($htmlarray['name']))
    {
        return 'error, one of the mandatory field of the function  select_sellist is missing';
    }
    $htmlName=$htmlarray['name'];
    $ajaxNbChar=$htmlarray['ajaxNbChar'];
    $listFields=explode(',',$sqlarray['fields']);
    $fields=array();
    foreach($listFields as $item){
        $start=MAX(strpos($item,' AS '),strpos($item,' as '));
        $label=($start)? substr($item, $start+4):$item;
        $fields[]=array('select' => $item, 'label'=>$label);
    }
    $select="\n";
    if ($ajax)
    {
        include_once DOL_DOCUMENT_ROOT . '/core/lib/ajax.lib.php';
       $token=getToken();
        $urloption='token='.$token;
        $comboenhancement = '';
        //$ajaxUrl='';
        $searchfields='';
        if($ajaxNbChar ){
            $ajaxUrl=dol_buildpath('/project_cost/core/ajaxGenericSelectHandler.php',1);
            $_SESSION['ajaxQuerry'][$token]['sql']=$sqlarray;
            $_SESSION['ajaxQuerry'][$token]['fields']=$fields;
            $_SESSION['ajaxQuerry'][$token]['html']=$htmlarray; 
            $_SESSION['ajaxQuerry'][$token]['option']=$addtionnalChoices;
            
                    //array('table'=>$table, 'fieldValue'=>$fieldValue,'htmlName'=> $htmlName,'fieldToShow1'=>$fieldToShow1,'fieldToShow2'=>$fieldToShow2,'separator'=> $separator,'sqlTailTable'=>$sqlTailTable,'sqlTailWhere'=>$sqlTailWhere,'addtionnalChoices'=>$addtionnalChoices);
            $comboenhancement = ajax_autocompleter($selected, $htmlName, $ajaxUrl, $urloption,$ajaxNbChar);
            $sqlTail.=" LIMIT 5";
            // put \\ before barket so the js will work for Htmlname before it is change to seatch HTMLname
            $htmlid=str_replace('[','\\\\[',str_replace(']','\\\\]',$htmlName));
            $comboenhancement=str_replace('#'.$htmlName, '#'.$htmlid,$comboenhancement);
            $comboenhancement=str_replace($htmlName.':', '"'.$htmlName.'":',$comboenhancement); // #htmlname doesn't cover everything
            $htmlName='search_'.$htmlName;
        }else{
            $comboenhancement = ajax_combobox($htmlName);
        }
        // put \\ before barket so the js will work
        $htmlid='#'.str_replace('[','\\\\[',str_replace(']','\\\\]',$htmlName));
        $comboenhancement=str_replace('#'.$htmlName, $htmlid,$comboenhancement);
        //incluse js code in the html response
        $select.=$comboenhancement;
        $nodatarole=($comboenhancement?' data-role="none"':'');
        
    }
    
    //dBQuerry
    $SelectOptions='';
    $selectedValue='';
    $sql='SELECT DISTINCT ';
    $sql.=$sqlarray['keyfield'];
    $sql.=' ,'.$sqlarray['fields'];
    $sql.= ' FROM '.MAIN_DB_PREFIX.$sqlarray['table'].' as t';
    if(isset($sqlarray['join']) && !empty($sqlarray['join']))
            $sql.=' '.$sqlarray['join'];
    if(isset($sqlarray['where']) && !empty($sqlarray['where']))
            $sql.=' WHERE '.$sqlarray['where'];
    if(isset($sqlarray['tail']) && !empty($sqlarray['tail']))
            $sql.=' '.$sqlarray['tail'];      
    dol_syslog('form::select_sellist ', LOG_DEBUG);
    
    $resql=$db->query($sql);
   
    if ($resql)
    {
        $selectOptions.= "<option value=\"-1\" ".(empty($selected)?"selected":"").">&nbsp;</option>\n";
        $i=0;
        $separator=isset($htmlarray['separator'])?$htmlarray['separator']:' ';
         //return $table."this->db".$field;
        $num = $db->num_rows($resql);
        while ($i < $num)
        {
            
            $obj = $db->fetch_object($resql);
            
            if ($obj)
            {
                $fieldtoshow='';
                foreach($fields as $item){
                    if(!empty($fieldtoshow))$fieldtoshow.=$separator;
                    $fieldtoshow.=$obj->{$label};
                }
                $selectOptions.= "<option value=\"".$obj->{$sqlarray['keyfield']}."\" ";
                if($obj->{$sqlarray['keyfield']}==$selected){
                     $selectOptions.='selected=\"selected\"';
                     $selectedValue=$fieldtoshow;
                 }
                 $selectOptions.=">";                    
                 $selectOptions.=$fieldtoshow;
                 $selectOptions.="</option>\n";
            } 
            $i++;
        }
        if($addtionnalChoices)foreach($addtionnalChoices as $value => $choice){
            $selectOptions.= '<option value="'.$value.'" '.(($selected==$value)?'selected':'').">{$choice}</option>\n";
        }
        
    }
    else
    {
        $error++;
        dol_print_error($db);
       $select.= "<option value=\"-1\" selected=\"selected\">ERROR</option>\n";
    }
     
   
    if($ajaxNbChar && $ajax){
        if ($selectedValue=='' && is_array($addtionnalChoices)){
            $selectedValue=$addtionnalChoices[$selected];
        }
        $select.='<input type="text" class="minwidth200 '.(isset($htmlarray['class'])?$htmlarray['class']:'').'" name="'.$htmlName.'" id="'.(isset($htmlarray['id'])?$htmlarray['id']:$htmlName).'" value="'.$selectedValue.'"'.$htmlarray['otherparam'].' />';
    }else{
        $select.='<select class="flat minwidth200 '.(isset($htmlarray['class'])?$htmlarray['class']:'').'" id="'.(isset($htmlarray['id'])?$htmlarray['id']:$htmlName).' name="'.$htmlName.'"'.$nodatarole.' '.$htmlarray['otherparam'].'>';
        $select.=$selectOptions;
        $select.="</select>\n";
    }
   // }
      return $select;

**Comment by delcroip** [comment]
I have done a select_sellist and an print_selllist function
 https://github.com/delcroip/dolibarr_page_from_SQLtable/blob/master/page_from_table/core/lib/generic.lib.php

the select is compatible with my generiqueAJaxhandler:
https://github.com/delcroip/dolibarr_page_from_SQLtable/blob/master/page_from_table/core/ajaxGenericSelectHandler.php
